core/shared/src/main/scala/scalaz/zio/internal/Env.scala

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) 2018 - 2019 John A. De Goes. All rights reserved.
+package scalaz.zio.internal
+
+import scalaz.zio._
+import scalaz.zio.ExitResult.Cause
+import java.util.concurrent.atomic.AtomicLong
+
+/**
+ * An environment provides the capability to execute different types
+ * of tasks.
+ */
+trait Env {
+
+  /**
+   * Retrieves the executor for the specified type of tasks.
+   */
+  def executor(tpe: Executor.Type): Executor
+
+  /**
+   * Retrieves the scheduler.
+   */
+  def scheduler: Scheduler
+
+  /**
+   * Determines if a throwable is non-fatal or not.
+   */
+  def nonFatal(t: Throwable): Boolean
+
+  /**
+   * Awaits for the result of the fiber to be computed.
+   */
+  final def unsafeRun[E, A](unhandled: Cause[Any] => IO[Nothing, _], io: IO[E, A]): A = {
+    val exit = unsafeRunSync(unhandled, io)
+
+    exit.fold(cause => throw new FiberFailure(cause), identity)
+  }
+
+  /**
+   * Awaits for the result of the fiber to be computed.
+   */
+  final def unsafeRunSync[E, A](unhandled: Cause[Any] => IO[Nothing, _], io: IO[E, A]): ExitResult[E, A] = {
+    val result = OneShot.make[ExitResult[E, A]]
+
+    unsafeRunAsync(unhandled, io, (x: ExitResult[E, A]) => result.set(x))
+
+    result.get
```

Comment:
no some kind of timeout?

Reply:
The `IO` can be timed out before being run. No need to duplicate that here.